torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2814,9 +2842,21 @@ def _fsdp_root_pre_forward(self, *args, **kwargs):
             return args, kwargs
         self._wait_for_previous_optim_step()
         self._needs_pre_forward_unshard.clear()
+        self._clear_grads_if_needed()
```

Comment:
add a comment to mention we are calling this to enable correctness when user has `set_grad_to_none` and it is a sort of delayed `set_to_none`.

Also, is it worth documenting these writeback semantics clearly to the end user?